inst/include/dust/filter.hpp

```diff
@@ -69,16 +100,86 @@ filter(T * obj,
       state.trajectories.advance();
     }
 
-    if (save_snapshots && state.snapshots.is_snapshot_step(d.first)) {
+    if (save_snapshots && state.snapshots.is_snapshot_step(step)) {
       obj->state_full(state.snapshots.value_iterator());
       state.snapshots.advance();
     }
+
+    if (step >= step_end) {
+      break;
+    }
   }
 
   return log_likelihood;
 }
 
 
+template <typename T>
+std::vector<typename T::real_type>
+deterministic(T * obj,
```

Comment:
Function looks very similar to `stochastic`. What were your thoughts vs a single function with a `deterministic` boolean and some if blocks?

Reply:
If keeping separate, then I think if any changes are made above (comment and loop init) they should also be done here

Reply:
these ones could actually be merged you're right (unlike the comment I left in mcstate